src/lib.rs

```diff
@@ -424,3 +489,22 @@ fn bytes_to_u32(bytes: &[u8], offset: &mut usize) -> Result<u32, ParseError> {
 fn shift_and_mask_lower(shift: u32, mask: u32, value: u32) -> u32 {
     (value >> shift) & ((1 << mask) - 1)
 }
+
+#[test]
+fn test_malformed_key_value_data_handling() {
+    let data = [
+        // Regular key-value pair
+        &7_u32.to_le_bytes()[..],
+        b"xyz\0123 ",
+        // Malformed key-value pair with missing NUL byte
+        &11_u32.to_le_bytes()[..],
+        b"abcdefghi!! ",
+        // Regular key-value pair again
+        &7_u32.to_le_bytes()[..],
+        b"xyz\0123",
+    ];
+
+    let iterator = KeyValueDataIterator { data: &data.concat() };
+
+    assert_eq!(iterator.count(), 2);
+}
```

Comment:
We should also have tests for bad lengths (`0`, `1000`, `u32::MAX`), and we should verify that correct pairs are returned.